File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/domain/projects_client.rb
# frozen_string_literal: true

module GreySdk
  module Domain
    # Projects client for project-related operations
    class ProjectsClient
      # @param http_client [GreySdk::Http::HttpClient] The HTTP client
      def initialize(http_client)
        @http = http_client
      end

      # Lists all projects for the current tenant
      # @param page [Integer, nil] Page number for pagination
      # @param per_page [Integer, nil] Number of items per page
      # @return [GreySdk::Error::Result] Result containing projects list or error
      def list_projects(page: nil, per_page: nil)
        params = {}
        params[:page] = page if page
        params[:per_page] = per_page if per_page

        @http.get("/api/v1/projects", params)
      end

      # Creates a new project
      # @param name [String] Project name
      # @param description [String, nil] Project description
      # @param metadata [Hash, nil] Additional project metadata
      # @return [GreySdk::Error::Result] Result containing created project or error
      def create_project(name:, description: nil, metadata: nil)
        return validation_error("Project name is required") if name.nil? || name.empty?

        body = { name: name }
        body[:description] = description if description
        body[:metadata] = metadata if metadata

        @http.post("/api/v1/projects", body)
      end

      # Retrieves a project by ID
      # @param project_id [String] The project ID
      # @return [GreySdk::Error::Result] Result containing project data or error
      def get_project(project_id:)
        return validation_error("Project ID is required") if project_id.nil? || project_id.to_s.empty?

        @http.get("/api/v1/projects/#{project_id}")
      end

      # Updates a project
      # @param project_id [String] The project ID
      # @param data [Hash] Project data to update
      # @return [GreySdk::Error::Result] Result containing updated project or error
      def update_project(project_id:, data:)
        return validation_error("Project ID is required") if project_id.nil? || project_id.to_s.empty?
        return validation_error("Update data is required") if data.nil? || data.empty?

        @http.patch("/api/v1/projects/#{project_id}", data)
      end

      # Deletes a project
      # @param project_id [String] The project ID
      # @return [GreySdk::Error::Result] Result indicating success or error
      def delete_project(project_id:)
        return validation_error("Project ID is required") if project_id.nil? || project_id.to_s.empty?

        @http.delete("/api/v1/projects/#{project_id}")
      end

      private

      def validation_error(message)
        Error::Result.err(Error::GreyError.validation(message))
      end
    end
  end
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-r/R/projects_client.R
#' @title Projects Client
#' @description Projects client for project-related operations
#' @name ProjectsClient
NULL

#' @importFrom R6 R6Class
#' @export
ProjectsClient <- R6::R6Class(
  "ProjectsClient",
  
  public = list(
    #' @description Create a new projects client
    #' @param http_client HttpClient instance
    initialize = function(http_client) {
      private$http <- http_client
    },
    
    #' @description List all projects
    #' @param page Page number (optional)
    #' @param per_page Items per page (optional)
    #' @return grey_result with projects list or error
    list_projects = function(page = NULL, per_page = NULL) {
      params <- list()
      if (!is.null(page)) params$page <- page
      if (!is.null(per_page)) params$per_page <- per_page
      
      if (length(params) == 0) params <- NULL
      
      private$http$get("/api/v1/projects", params)
    },
    
    #' @description Create a new project
    #' @param name Project name
    #' @param description Project description (optional)
    #' @param metadata Additional metadata (optional)
    #' @return grey_result with created project or error
    create_project = function(name, description = NULL, metadata = NULL) {
      if (is.null(name) || nchar(trimws(name)) == 0) {
        return(result_err(error_validation("Project name is required")))
      }
      
      body <- list(name = name)
      if (!is.null(description)) body$description <- description
      if (!is.null(metadata)) body$metadata <- metadata
      
      private$http$post("/api/v1/projects", body)
    },
    
    #' @description Get a project by ID
    #' @param project_id The project ID
    #' @return grey_result with project data or error
    get_project = function(project_id) {
      if (is.null(project_id) || nchar(trimws(as.character(project_id))) == 0) {
        return(result_err(error_validation("Project ID is required")))
      }
      
      private$http$get(paste0("/api/v1/projects/", project_id))
    },
    
    #' @description Update a project
    #' @param project_id The project ID
    #' @param data Update data (list)
    #' @return grey_result with updated project or error
    update_project = function(project_id, data) {
      if (is.null(project_id) || nchar(trimws(as.character(project_id))) == 0) {
        return(result_err(error_validation("Project ID is required")))
      }
      
      if (is.null(data) || length(data) == 0) {
        return(result_err(error_validation("Update data is required")))
      }
      
      private$http$patch(paste0("/api/v1/projects/", project_id), data)
    },
    
    #' @description Delete a project
    #' @param project_id The project ID
    #' @return grey_result
    delete_project = function(project_id) {
      if (is.null(project_id) || nchar(trimws(as.character(project_id))) == 0) {
        return(result_err(error_validation("Project ID is required")))
      }
      
      private$http$delete(paste0("/api/v1/projects/", project_id))
    }
  ),
  
  private = list(
    http = NULL
  )
)
